src/blockdata/locktime/relative.rs

```diff
@@ -0,0 +1,155 @@
+// Rust Bitcoin Library - Written by the rust-bitcoin developers.
+// SPDX-License-Identifier: CC0-1.0
+
+//! Provides type [`LockTime`] that implements the logic around nSequence/OP_CHECKSEQUENCEVERIFY.
+//!
+//! There are two types of lock time: lock-by-blockheight and lock-by-blocktime, distinguished by
+//! whether bit 22 of the `u32` consensus value is set.
+//!
+
+use crate::blockdata::transaction::Sequence;
+
+#[cfg(docsrs)]
+use crate::relative;
+
+/// A relative lock time value, representing either a block height or time (512 second intervals).
+///
+/// The `relative::LockTime` type does not have any constructors, this is by design, please use
+/// `Sequence::to_relative_lock_time` to create a relative lock time.
+///
+/// ### Relevant BIPs
+///
+/// * [BIP 68 Relative lock-time using consensus-enforced sequence numbers](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki)
+/// * [BIP 112 CHECKSEQUENCEVERIFY](https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki)
+#[allow(clippy::derive_ord_xor_partial_ord)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub enum LockTime {
+    /// A block height lock time value.
+    Blocks(Height),
+    /// A 512 second time interval value.
+    Time(Time),
+}
+
+impl LockTime {
+    /// Creates a [`relative::LockTime`], locked by height, from `value`.
+    pub(crate) fn from_height(value: u16) -> Self {
+        LockTime::Blocks(Height(value))
+    }
+
+    /// Creates a [`relative::LockTime`], locked by time intervals, from `value`.
+    pub(crate) fn from_time(value: u16) -> Self {
+        LockTime::Time(Time(value))
+    }
+
+    /// Returns true if this [relative::`LockTime`] is satisfied by sequence number `n`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use bitcoin::{relative, Sequence};
+    ///
+    /// let height: u16 = 100;
+    /// let lock = Sequence::from_height(height).to_relative_lock_time().expect("valid height");
+    /// assert!(lock.is_satisfied_by(Sequence::from_height(height+1)));
+    /// ```
+    // FIXME: Have I got the terminology around the correct way? Does self satisfy n or n satisfy self?
+    pub fn is_satisfied_by(&self, n: Sequence) -> bool {
```

Comment:
Oh, if you keep the existing `false` returns (see below) then `&&` should be replaced by `||`.

Reply:
I'm still not fully understanding where the height and time values would come from to pass into this new function. Would a user of this function have access to the chain and be expected to be able to compute the current height and time?  I've based this (and the `absolute::LockTime` work) off of the usage in `miniscript`, is it that `miniscript` is not a normal user? How would we implement `check_older` in `miniscript` if we cannot use a `Sequence` number to satisfy the relative lock time?

Reply:
Ah, I understand the question now. So, there are two pieces to locktime validation in Bitcoin:

1. The `CHECKSEQUENCEVERIFY` (and `CHECKLOCKTIMEVERIFY`) opcodes, which check the sequence/locktime values in a transaction against some "test" sequence/locktime value, and decides whether or not it passes. This is what rust-miniscript uses when reasoning about `check_older`/`check_after`.
2. **Separately** the Bitcoin blockchain checks whether the sequence/locktime values actually indicate a valid locktime, based on the state of the chain. This is what I think "normal users" are thinking of when they work with sequence/locktime values.

I would suggest that a "normal user" is more likely to be in situation (2), whereas situation (1) corresponds to Script interpretation. In this API we should somehow support both usecases somehow

Reply:
I think that my suggested API, in which `is_satisfied_by` takes both a height and time, would serve both cases.

In case (1), rust-miniscript will see `check_older(N)`, decide whether to interpret `N` as a height or time, and then use `is_satisfied_by_height`/`is_satisfied_by_time` as appropriate. In case (2) the user has both the current height and current time, and can use `is_satisfied_by`.

Miniscript *additionally* needs a way to serialize a height/time as a "sequence number" to embed into Script, and to parse out such a value.

Reply:
Thanks man! Your suggested API works quite nicely, here is `check_older` from `miniscript`
```rust
     match self.to_relative_lock_time() {
            None => false,
            Some(seq) => match (n, seq) {
                (Blocks(_), Blocks(h)) => n.is_satisfied_by_height(h.value()),
                (Time(_), Time(t)) => n.is_satisfied_by_time(t.value()),
                _ => false, // Not the same units.
            }
        }

```

Reply:
Although it could equally have been written without using the `is_satisfied_by_*` methods
```rust
        match self.to_relative_lock_time() {
            None => false,
            Some(seq) => match (n, seq) {
                (Blocks(n), Blocks(lock_time)) => n.value() <= lock_time.value(),
                (Time(n), Time(lock_time)) => n.value() <= lock_time.value(),
                _ => false, // Not the same units.
            }
        }
```

I was only matching against `self` and tests were failing until I matched against both `self` and `n`, so the API is not bullet proof.

Reply:
It probably can't be helped, unless you pull that entire block of code into its own API function here :)

Reply:
Idea how to resolve having to come up with different names:

```rust
// maybe this doesn't have to be sealed, IDK
pub trait SatisfiesRelativeLockTime: sealed::SatisfiesRelativeLockTime {
    fn satisfies_relative_lock_time(self, lock_time: relative::LockTime) -> bool;
}

impl SatisfiesRelativeLockTime for Sequence { /* ... */ }
impl SatisfiesRelativeLockTime for (Height, Time) { /* ... */ }
// impl SatisfiesRelativeLockTime for (Time, Height) { /* ... */ } // for convenience so people don't have to remember the order?

impl relative::LockTime {
    pub fn is_satisfied_by<T: SatisfiesRelativeLockTime>(self, satisified_by: T) -> bool {
        satisfied_by.satisfies_relative_lock_time(self)
    }
}
```

Reply:
I like the trait idea. Implementing for `Sequence` goes against what @apoelstra says above (quoted below)

> I don't think it makes sense for a locktime to be satisfied by a sequence no, or vice-versa. Rather, I would expect a sequence number to be convertable to a relative::Locktime, and then this would be satisfied by a blockheight and/or timestamp.

> I don't think it makes sense for a locktime to be satisfied by a sequence no, or vice-versa. Rather, I would expect a sequence number to be convertable to a relative::Locktime, and then this would be satisfied by a blockheight and/or timestamp.

If `Sequence` disables lock time it is not possible to construct a `relative::LockTime` from the sequence so this concern is addressed. I think we should implement for `relative::LockTime`, `(Height, Time)`, and   `(Time, Height)`.

Reply:
I implemented for `Height` and `Time` as well as for the tuple of both.

Reply:
> If `Sequence` disables lock time it is not possible to construct a `relative::LockTime` from the sequence so this concern is addressed.

I still didn't like the possibility of manual check but now I think maybe that's too much paranoia.

Reply:
Oh, crap, I now realized it should've been the other way around. It makes sense to check if sequence is satisfied by `(Height, Time)`|`RelativeLockTime` but not that much for relative lock time.